harness/determined/pytorch/_trainer.py

```diff
@@ -0,0 +1,212 @@
+
+            self._trial_controller = pytorch.PyTorchTrialController(
+                trial_inst=self._trial,
+                context=self._context,
+                max_length=max_length,
+                min_validation_period=min_validation_period,
+                min_checkpoint_period=min_checkpoint_period,
+                average_training_metrics=average_training_metrics,
+                checkpoint_policy=checkpoint_policy,
+                smaller_is_better=smaller_is_better,
+                local_training=True,
+            )
+        else:
+            self._trial_controller = pytorch.PyTorchTrialController(
+                trial_inst=self._trial,
+                context=self._context,
+                min_checkpoint_period=min_checkpoint_period,
+                min_validation_period=min_validation_period,
+                average_training_metrics=average_training_metrics,
+                checkpoint_policy=checkpoint_policy,
+                latest_checkpoint=self._cluster_info.latest_checkpoint,
+                smaller_is_better=smaller_is_better,
+                searcher_metric_name=self._cluster_info.trial._config["searcher"]["metric"],
+                local_training=False,
+                det_profiler=self._det_profiler,
+                steps_completed=self._cluster_info.trial._steps_completed,
+            )
+
+        self._trial_controller.run()
+
+    def _convert_period_to_train_unit(self, period: int, train_unit: pytorch.TrainUnit):
+        # Local training will assume same period as max_length
+        if self._local_training:
+            if isinstance(train_unit, pytorch.Batch):
+                return pytorch.Batch(period)
+            elif isinstance(train_unit, pytorch.Epoch):
+                return pytorch.Epoch(period)
+            elif isinstance(train_unit, pytorch.Record):
+                return pytorch.Record(period)
+
+        # On-cluster training assumes period lengths with the searcher
+        searcher_unit = self._core.searcher.get_configured_units()
+        return pytorch.TrainUnit._from_searcher_unit(period, searcher_unit)
+
+
+def _initialize_distributed_backend():
+    distributed_backend = det._DistributedBackend()
+    if distributed_backend.use_horovod():
+        hvd.require_horovod_type("torch", "PyTorchTrial is in use.")
+        hvd.init()
+        return core.DistributedContext.from_horovod(horovod.hvd)
+    elif distributed_backend.use_torch():
+        if torch.cuda.is_available():
+            dist.init_process_group(backend="nccl")  # type: ignore
+        else:
+            dist.init_process_group(backend="gloo")  # type: ignore
+        return core.DistributedContext.from_torch_distributed()
+    else:
+        print(f"Backend {distributed_backend} not supported")
+
+
+def _generate_local_seed():
+    return random.randint(0, 1 << 31)
+
+
+@contextlib.contextmanager
+def init(
+    hparams: Optional[Dict] = None, distributed_context: Optional[core.DistributedContext] = None
```

Comment:
I have a slight preference for making these kwarg-only (by including an initial `*` arg)

Also, please rename `distributed_context` to just `distributed` to match `det.core.init()`.

Reply:
renamed. but why do you want to make these kwarg only? this init only supports a set number of args to be passed to PyTorchContext constructor. isn't it better to be explicit about what is configurable?

Reply:
I'm not saying to accept `**kwarg`, I'm saying make the existing args kwarg only:
```python
def init(*, hparams: Optional[dict] = None, distributed_context: Optional[core.DistributedContext] = None):
```

It makes our APIs a little more flexible for us to change in the future.

It's less important with `det.pytorch.init` than with `det.core.init` because I think `core.init` is more likely to get new args added, but I like the consistency of it.